src/main.cpp

```diff
@@ -9,7 +9,38 @@
 int main(int argc, char** argv) {
   try {
     LightGBM::Application app(argc, argv);
-    app.Run();
+
+    std::string exception_str;
+    try
+    {
+      app.Run();
+    }
+    catch (const std::exception& ex) {
+      exception_str = ex.what();
+    }
+    catch (const std::string& ex) {
+      exception_str = ex;
+    }
+    catch (...) {
+      exception_str = "Unknown Exceptions";
+    }
+
+    if (!exception_str.empty())
+    {
+#ifdef USE_MPI
+      // In the MPI mode, if there is an exception and you try to exit this block, the destructor for "app" is called,
+      // which in turn calls MPI_Finalize(). But that causes this process to hang and the whole MPI job can hang as a result.
+      // Instead we may want to call MPI_Abort(). However that right now doesn't do much more call abort(). So we just do it here.
+      if (app.IsParallel())
+      {
+        std::cerr << "Aborting because of exception in MPI mode:" << std::endl;
+        std::cerr << exception_str << std::endl;
+        std::cerr.flush();
+        abort();
+      }
+#endif
+      throw exception_str;
```

Comment:
Thanks so much! Indeed the destructor will cause the problem. 
However, I think the two-level `try ... catch` may is not a good fix. 
I think a possible solution to move the `MPI_Finalize` out of destructor, and call `MPI_Abort` if any exception met, otherwise call the `MPI_Finalize`.